### PR TITLE
Fix unzip on win php56

### PR DESCRIPTION
**Pull request by gza**
Refactoring of unzip ZipArchive fallback.

* a solution to #604
* cleaner fallback mecanism between ZipArchive and SystemUnZip
* tests

### CODE REVIEW

File: src/Composer/Downloader/ZipDownloader.php [hunk 1]
```diff
@@ -59,42 +64,112 @@ public function download(PackageInterface $package, $path, $output = true)
...
         }
+        return new \RuntimeException($processError);
+    }
 
+    /**
+     * extract $file to $path with ZipArchive
+     *
+     * @param string $file File to extract
+     * @param string $path Path where to extract file
+     * @return bool True if succeed
```
Comment (stof) [inline src/Composer/Downloader/ZipDownloader.php]: that's totally wrong. Your method does not return a boolean. It returns ``bool|\Exception`` (which is a very weird signature btw)

File: src/Composer/Downloader/ZipDownloader.php [hunk 2]
```diff
@@ -59,42 +64,112 @@ public function download(PackageInterface $package, $path, $output = true)
...
     }
 
-    protected function extract($file, $path)
+    /**
+     * extract $file to $path with "unzip" command
+     *
+     * @param string $file      File to extract
+     * @param string $path      Path where to extract file
+     * @param bool $isFallback  If true it is called as a fallback and should not throw exception
+     * @return bool True if succeed
```
Comment (stof) [inline src/Composer/Downloader/ZipDownloader.php]: wrong. It returns ``bool|\Exception|null``

File: src/Composer/Downloader/ZipDownloader.php [hunk 3]
```diff
@@ -59,42 +64,112 @@ public function download(PackageInterface $package, $path, $output = true)
...
+     */
+    protected function extractWithSystemUnzip($file, $path, $isFallback)
```
Comment (stof) [inline src/Composer/Downloader/ZipDownloader.php]: protected methods should go after public ones.

and btw, why are they protected ?

File: src/Composer/Downloader/ZipDownloader.php [hunk 4]
```diff
@@ -59,42 +64,111 @@ public function download(PackageInterface $package, $path, $output = true)
...
+        return TRUE;
+    }
+
+    /**
+     * extract $file to $path
+     *
+     * @param string $file File to extract
+     * @param string $path Path where to extract file
+     */
+    public function extract($file, $path)
```
Comment (stof) [inline src/Composer/Downloader/ZipDownloader.php]: why public ? This method should be protected, according to the parent requirement

File: src/Composer/Downloader/ZipDownloader.php (hunk 3 above)
Comment (gza) [inline src/Composer/Downloader/ZipDownloader.php]: I refactored a big "protected" function into 3 ones.
I didn't challenged that choice and created 3 protected functions.

File: src/Composer/Downloader/ZipDownloader.php (hunk 4 above)
Comment (gza) [inline src/Composer/Downloader/ZipDownloader.php]: ooops, protected was a problem during testing, but I worked around it.
I should not have commited this change.

thanks !

**Comment by gza** [comment]
should I switch target branch to 1.3 ?

**Comment by Seldaek** [comment]
I am not entirely sure what this fixes.. Right now on windows we use ZipArchive if present, and otherwise unzip if present. On linux/osx we use unzip if present, otherwise ZipArchive (the order is reversed because unzip is faster IIRC, but on windows it tends to create problems).

Your PR changes the order to always be ZipArchive if present then unzip. Why? How is that fixing anything?

**Comment by gza** [comment]
Yes, when ZipArchive fails (like in #604), its falls back to unzip

**Comment by Seldaek** [comment]
Yeah I can see that.. but I'm asking why / what you are trying to achieve with this PR? I don't see why I would merge it right now.

**Comment by gza** [comment]
Ooops, excuse me, I didn't saw I was changing unzip/pecl priority on unix. I was focus on my windows problem.

I should restore original unix behavour by switching order priority depending on Platform::isWindows().
But, I need to implement fallback on ZipArchive errors.

Is it good for you ?

**Comment by Seldaek** [comment]
Well no sorry but that still does not explain what your PR is fixing at all? I don't care about the code right now, I'm trying to understand the intent/goal of the PR.

**Comment by gza** [comment]
this is workaround to an issue that appends with:
* PHP < 7.1
* Windows
* Composer packages with long pathes (like zendframework1)

**Comment by Seldaek** [comment]
OK but where is the workaround? As far as I can tell, we currently try ZipArchive and then unzip on windows. And after your PR it does that too but for linux as well.. so what did you effectively change to fix that apart from refactoring the code a bit?

**Comment by gza** [comment]
Original code does never fallback on windows, its just never does unzip on windows : 
https://github.com/composer/composer/blob/master/src/Composer/Downloader/ZipDownloader.php#L66

**Comment by Seldaek** [comment]
Ah right ok yeah I guess I misread that the other day. So yes if you can update it to use unzip>ZipArchive on linux and the opposite on windows that would be great and then I think it's good to merge :)

**Comment by gza** [comment]
thanks, I will do it tomorrow :)

File: src/Composer/Downloader/ZipDownloader.php [hunk 5]
```diff
@@ -59,42 +64,111 @@ public function download(PackageInterface $package, $path, $output = true)
...
+    protected function extractWithZipArchive($file, $path)
+    {
         $zipArchive = new ZipArchive();
 
         if (true !== ($retval = $zipArchive->open($file))) {
-            throw new \UnexpectedValueException(rtrim($this->getErrorMessage($retval, $file)."\n".$processError), $retval);
+            return new \UnexpectedValueException(rtrim($this->getErrorMessage($retval, $file)."\n"), $retval);
+        }
+
+        $extractResult = FALSE;
```
Comment (povils) [inline src/Composer/Downloader/ZipDownloader.php]: null, false, true.. is in lowercase

File: src/Composer/Downloader/ZipDownloader.php [hunk 6]
```diff
@@ -59,42 +64,111 @@ public function download(PackageInterface $package, $path, $output = true)
...
 
-        if (true !== $zipArchive->extractTo($path)) {
-            throw new \RuntimeException(rtrim("There was an error extracting the ZIP file, it is either corrupted or using an invalid format.\n".$processError));
+        if (true !== $extractResult) {
+            return new \RuntimeException(rtrim("There was an error extracting the ZIP file, it is either corrupted or using an invalid format.\n"));
         }
 
         $zipArchive->close();
+
+        return TRUE;
```
Comment (povils) [inline src/Composer/Downloader/ZipDownloader.php]: null, false, true.. is in lowercase

File: src/Composer/Downloader/ZipDownloader.php [hunk 7]
```diff
@@ -59,42 +64,111 @@ public function download(PackageInterface $package, $path, $output = true)
...
+
+    /**
+     * extract $file to $path
+     *
+     * @param string $file File to extract
+     * @param string $path Path where to extract file
+     */
+    protected function extract($file, $path)
+    {
+        $resultZipArchive = NULL;
```
Comment (povils) [inline src/Composer/Downloader/ZipDownloader.php]: null, false, true.. is in lowercase

File: src/Composer/Downloader/ZipDownloader.php [hunk 8]
```diff
@@ -59,42 +64,111 @@ public function download(PackageInterface $package, $path, $output = true)
...
+        $resultUnzip = NULL;
```
Comment (povils) [inline src/Composer/Downloader/ZipDownloader.php]: null, false, true.. is in lowercase

File: src/Composer/Downloader/ZipDownloader.php [hunk 9]
```diff
@@ -59,42 +64,111 @@ public function download(PackageInterface $package, $path, $output = true)
...
+
+        if ( self::$hasZipArchive ) {
+            // zip module is present
+            $resultZipArchive = $this->extractWithZipArchive($file, $path);
+            if ($resultZipArchive === TRUE) {
```
Comment (povils) [inline src/Composer/Downloader/ZipDownloader.php]: null, false, true.. is in lowercase

File: src/Composer/Downloader/ZipDownloader.php [hunk 10]
```diff
@@ -59,42 +64,111 @@ public function download(PackageInterface $package, $path, $output = true)
...
+                return;
+            }
+        }
+
+        if ( self::$hasSystemUnzip ) {
+            // we have unzip in the path
+            $isFallback=FALSE;
```
Comment (povils) [inline src/Composer/Downloader/ZipDownloader.php]: null, false, true.. is in lowercase. ALso spaces

File: src/Composer/Downloader/ZipDownloader.php [hunk 11]
```diff
@@ -59,42 +64,111 @@ public function download(PackageInterface $package, $path, $output = true)
...
+
+        if ( self::$hasZipArchive ) {
+            // zip module is present
+            $resultZipArchive = $this->extractWithZipArchive($file, $path);
+            if ($resultZipArchive === TRUE) {
+                return;
+            }
+        }
+
+        if ( self::$hasSystemUnzip ) {
```
Comment (povils) [inline src/Composer/Downloader/ZipDownloader.php]: why spaces around?

File: src/Composer/Downloader/ZipDownloader.php [hunk 12]
```diff
@@ -59,42 +64,111 @@ public function download(PackageInterface $package, $path, $output = true)
...
+     *
+     * @param string $file File to extract
+     * @param string $path Path where to extract file
+     */
+    protected function extract($file, $path)
+    {
+        $resultZipArchive = NULL;
+        $resultUnzip = NULL;
+
+        if ( self::$hasZipArchive ) {
```
Comment (povils) [inline src/Composer/Downloader/ZipDownloader.php]: why these spaces?

File: src/Composer/Downloader/ZipDownloader.php [hunk 13]
```diff
@@ -59,42 +64,111 @@ public function download(PackageInterface $package, $path, $output = true)
...
+
+        if ( self::$hasSystemUnzip ) {
+            // we have unzip in the path
+            $isFallback=FALSE;
+            if ( $resultZipArchive !== NULL) {
+                $this->io->writeError("\nUnzip using ZipArchive failed, trying with unzip");
+                $isFallback=TRUE;
+            };
+            $resultUnzip = $this->extractWithSystemUnzip($file, $path, $isFallback);
+            if ( $resultUnzip === TRUE ) {
```
Comment (povils) [inline src/Composer/Downloader/ZipDownloader.php]: why these spaces?

File: src/Composer/Downloader/ZipDownloader.php [hunk 14]
```diff
@@ -59,42 +64,111 @@ public function download(PackageInterface $package, $path, $output = true)
...
+            $resultZipArchive = $this->extractWithZipArchive($file, $path);
+            if ($resultZipArchive === TRUE) {
+                return;
+            }
+        }
+
+        if ( self::$hasSystemUnzip ) {
+            // we have unzip in the path
+            $isFallback=FALSE;
+            if ( $resultZipArchive !== NULL) {
```
Comment (povils) [inline src/Composer/Downloader/ZipDownloader.php]: why these spaces?

File: src/Composer/Downloader/ZipDownloader.php [hunk 15]
```diff
@@ -59,42 +64,111 @@ public function download(PackageInterface $package, $path, $output = true)
...
+                $this->io->writeError("\nUnzip using ZipArchive failed, trying with unzip");
+                $isFallback=TRUE;
```
Comment (povils) [inline src/Composer/Downloader/ZipDownloader.php]: null, false, true.. is in lowercase

File: src/Composer/Downloader/ZipDownloader.php (hunk 13 above)
Comment (povils) [inline src/Composer/Downloader/ZipDownloader.php]: why not just 
```php
if ($resultUnzip) {
```

**Comment by gza** [comment]
@povils : I'm sorry I will not respond to each of your comments, but I have noted to run some phpcs-fixer on the final code. Thanks

btw: if ($resultUnzip) is not equivalent to if ( $resultUnzip === TRUE )

**Comment by Seldaek** [comment]
Yeah the minor CS considerations are not a huge deal really, I run php-cs-fixer every now and then to clean things up myself. And commenting on every single one just spams people needlessly, please don't do this @povils.

**Comment by povils** [comment]
@gza yes it's not equivalent if $resultUnzip is not strictly boolean
@Seldaek sorry bad habits :)

Regards

**Comment by gza** [comment]
Well, I think it's ready ( and CS fixed ;-) )

**Comment by Seldaek** [comment]
Thanks :)

**Pull request merged**